src/dbnode/encoding/proto/buffer.go

```diff
@@ -0,0 +1,291 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// This file is mostly copy-pasta from: https://github.com/jhump/protoreflect/blob/master/dynamic/codec.go
+// since the jhump/protoreflect library does not expose the `codedBuffer` type.
+
+package proto
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"math"
+)
+
+var (
+	// errOverflow is returned when an integer is too large to be represented.
+	errOverflow = errors.New("proto: integer overflow")
+)
+
+// A reader/writer type that assists with decoding protobuf's binary representation.
+// This code is largely a fork of proto.Buffer, which cannot be used because it has no exported
+// field or method that provides access to its underlying reader index.
+type buffer struct {
+	buf   []byte
+	index int
+}
+
+func newCodedBuffer(buf []byte) *buffer {
+	return &buffer{buf: buf}
+}
+
+func (cb *buffer) reset(b []byte) {
+	cb.buf = b
+	cb.index = 0
+}
+
+func (cb *buffer) eof() bool {
+	return cb.index >= len(cb.buf)
+}
+
+func (cb *buffer) skip(count int) (int, bool) {
+	newIndex := cb.index + count
+	if newIndex > len(cb.buf) {
+		return 0, false
+	}
+	cb.index = newIndex
+	return 0, true
+}
+
+func (cb *buffer) decodeVarintSlow() (x uint64, err error) {
+	i := cb.index
+	l := len(cb.buf)
+
+	for shift := uint(0); shift < 64; shift += 7 {
+		if i >= l {
+			err = io.ErrUnexpectedEOF
+			return
+		}
+		b := cb.buf[i]
+		i++
+		x |= (uint64(b) & 0x7F) << shift
+		if b < 0x80 {
+			cb.index = i
+			return
+		}
+	}
+
+	// The number is too large to represent in a 64-bit value.
+	err = errOverflow
+	return
+}
+
+// DecodeVarint reads a varint-encoded integer from the Buffer.
+// This is the format for the
+// int32, int64, uint32, uint64, bool, and enum
+// protocol buffer types.
+func (cb *buffer) decodeVarint() (uint64, error) {
```

Comment:
the bit management for these functions looks sane, but i'm on the fence as to whether we should pull some of this business out into more testable components - right now, branch coverage is entirely dependent on input properties, which implicitly bloats the tests and makes test cases the driver rather than tests themselves. (generally, table-based tests test low-level variations on a single theme versus a gamut of functionality.)

tbh, i think that, at a minimum, we should test the buffer directly, but ideally we'd also have discrete tests for each type decoder.

what do you think?

Reply:
same comment as above, this is actualy copy-pasta from a library and hopefully a P.R will land upstream soon that means I don't have to keep this code in M3DB. Also, the buffer is not tested directly in that library so I can't even copy paste the tests :p